vault/src/vaults.rs

```diff
@@ -69,26 +76,42 @@ impl<P: RelayPallet + BtcRelayPallet, B: BitcoinCoreApi + Clone> VaultTheftMonit
     async fn report_invalid(
         &self,
         vault_id: &VaultId,
-        proof: &[u8],
-        raw_tx: &[u8],
         transaction: &Transaction,
+        block_hash: BlockHash,
     ) -> Result<(), Error> {
-        tracing::info!("Found tx from vault {}", vault_id.pretty_printed());
+        tracing::debug!(
+            "Found txid {} from vault {}",
+            transaction.txid(),
+            vault_id.pretty_printed()
+        );
         // check if matching redeem or replace request
-        if self.btc_parachain.is_transaction_invalid(vault_id, raw_tx).await? {
-            tracing::info!("Transaction is invalid");
-            self.btc_parachain.report_vault_theft(vault_id, proof, raw_tx).await?;
+        let raw_tx = bitcoin::serialize(transaction); //self.bitcoin_core.get_raw_tx(&tx_id, &block_hash).await?;
+        if self.btc_parachain.is_transaction_invalid(vault_id, &raw_tx).await? {
+            tracing::info!(
+                "Detected theft by vault {} - txid {}. Reporting...",
+                vault_id.pretty_printed(),
+                transaction.txid()
+            );
+
+            let proof = self.bitcoin_core.get_proof(transaction.txid(), &block_hash).await?;
+            self.btc_parachain.report_vault_theft(vault_id, &proof, &raw_tx).await?;
         } else {
             // valid payment.. but check that it is not a duplicate payment
-            for (txid_2, block_hash) in self.bitcoin_core.find_duplicate_payments(transaction).await? {
-                let raw_tx_2 = self.bitcoin_core.get_raw_tx(&txid_2, &block_hash).await?;
-                let proof_2 = self.bitcoin_core.get_proof(txid_2, &block_hash).await?;
+            for (txid_2, block_hash_2) in self.bitcoin_core.find_duplicate_payments(transaction).await? {
+                let raw_tx_2 = self.bitcoin_core.get_raw_tx(&txid_2, &block_hash_2).await?;
                 if !self.btc_parachain.is_transaction_invalid(vault_id, &raw_tx_2).await? {
-                    tracing::info!("Found a double payment - reporting...");
+                    tracing::info!(
+                        "Detected double payment by vault {} - txids {} and {}. Reporting...",
+                        vault_id.pretty_printed(),
+                        transaction.txid(),
+                        txid_2
+                    );
+                    let proof_1 = self.bitcoin_core.get_proof(transaction.txid(), &block_hash).await?;
+                    let proof_2 = self.bitcoin_core.get_proof(txid_2, &block_hash_2).await?;
```

Comment:
Probably doesn't make sense to parallelize since underlying implementation is synchronous but maybe we can leave a comment here?

Reply:
> Probably doesn't make sense to parallelize since underlying implementation is synchronous but maybe we can leave a comment here?

I don't think there's a need - if there are so many double payments that this becomes a bottleneck, then we have a bigger problem on our hands